apollo-federation/src/sources/connect/header.rs

```diff
@@ -0,0 +1,279 @@
+//! Headers defined in connectors `@source` directives.
+use std::str::FromStr;
+
+use nom::branch::alt;
+use nom::bytes::complete::tag;
+use nom::character::complete::alpha1;
+use nom::character::complete::alphanumeric1;
+use nom::character::complete::char;
+use nom::character::complete::none_of;
+use nom::combinator::map;
+use nom::combinator::recognize;
+use nom::multi::many0;
+use nom::multi::many1;
+use nom::sequence::delimited;
+use nom::sequence::pair;
+use nom::IResult;
+use serde_json_bytes::ByteString;
+use serde_json_bytes::Map;
+use serde_json_bytes::Value as JSON;
+
+/// A header value, optionally containing variable references.
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct HeaderValue {
+    parts: Vec<HeaderValuePart>,
+}
+
+impl HeaderValue {
+    fn new(parts: Vec<HeaderValuePart>) -> Self {
+        Self { parts }
+    }
+
+    fn parse(input: &str) -> IResult<&str, Self> {
+        map(many1(HeaderValuePart::parse), Self::new)(input)
+    }
+
+    /// Replace variable references in the header value with the given variable definitions.
+    ///
+    /// # Errors
+    /// Returns an error if a variable used in the header value is not defined or if a variable
+    /// value is not a string.
+    pub fn interpolate(&self, vars: &Map<ByteString, JSON>) -> Result<String, String> {
```

Comment:
what happens if you try to interpolate an object? like
```yaml
$config:
  foo:
    bar: 1
```
with
```graphql
headers: [{ name: "x-foo" value: "$config.foo" }] # not $config.foo.bar
```

Reply:
we can't validate against that because composition isn't aware of router config, so i'm curious what the error message is. 

maybe some unit tests for interpolation here to help cover that?

Reply:
The error is: `Unable to interpolate header value: "Variable $config.foo is not a string"`, and the header won't be included in the request. I'll add a unit test.

Reply:
i dunno why my eyes completely glazed over that a few lines down! what if the value is a JSON number? i can't remember if it's stringified

Reply:
or bool, i guess

Reply:
I added code to handle all the types, and unit tests to test.